In [1]:
import numpy as np
import pandas as pd
import re
import datetime
pd.set_option('display.max_columns', 500)

In [2]:
df = pd.read_csv('data/listings.csv')

C:\Users\georg\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (61,62,95) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


count      20769
unique       521
top       $50.00
freq         803
Name: price, dtype: object

In [7]:
df_missing.shape

(106, 2)

In [11]:
size = df.shape[0]
missing_cnt = df.isnull().sum()
missing_pct = missing_cnt/size*100
df_missing = pd.concat([missing_cnt,missing_pct],axis=1)
df_missing.columns = ['Missing Count','Missing Percentage']
df_missing[df_missing['Missing Count']>0].sort_values('Missing Count', ascending=False)

Missing Count  Missing Percentage
thumbnail_url                         20769          100.000000
medium_url                            20769          100.000000
host_acceptance_rate                  20769          100.000000
neighbourhood_group_cleansed          20769          100.000000
xl_picture_url                        20769          100.000000
jurisdiction_names                    20767           99.990370
license                               20765           99.980741
square_feet                           20613           99.248881
monthly_price                         18991           91.439164
weekly_price                          18682           89.951370
notes                                 10969           52.814291
host_about                             8758           42.168617
access                                 7986           38.451538
interaction                            7716           37.151524
neighborhood_overview                  7293           35.114835
transit                                7087           34.122972
house_rules                            6577           31.667389
space                                  5774           27.801050
host_response_time                     4984           23.997304
host_response_rate                     4984           23.997304
security_deposit                       4902           23.602484
review_scores_location                 4288           20.646155
review_scores_value                    4285           20.631711
review_scores_checkin                  4283           20.622081
review_scores_accuracy                 4282           20.617266
review_scores_communication            4280           20.607636
review_scores_cleanliness              4280           20.607636
review_scores_rating                   4272           20.569117
last_review                            3983           19.177620
reviews_per_month                      3983           19.177620
first_review                           3983           19.177620
cleaning_fee                           3384           16.293514
host_neighbourhood                     2520           12.133468
summary                                 664            3.197073
zipcode                                 365            1.757427
description                             346            1.665944
market                                   37            0.178150
state                                    28            0.134816
beds                                     23            0.110742
host_location                            19            0.091482
bathrooms                                15            0.072223
bedrooms                                  8            0.038519
host_identity_verified                    5            0.024074
host_has_profile_pic                      5            0.024074
host_total_listings_count                 5            0.024074
host_listings_count                       5            0.024074
host_picture_url                          5            0.024074
host_thumbnail_url                        5            0.024074
host_is_superhost                         5            0.024074
host_since                                5            0.024074
host_name                                 5            0.024074
city                                      1            0.004815
neighbourhood                             1            0.004815
name                                      1            0.004815

In [22]:
df.head(3)

id                         listing_url       scrape_id last_scraped  \
0   1419   https://www.airbnb.com/rooms/1419  20190604202502   2019-06-05   
1   8077   https://www.airbnb.com/rooms/8077  20190604202502   2019-06-06   
2  12604  https://www.airbnb.com/rooms/12604  20190604202502   2019-06-05   

                                 name  \
0     Beautiful home in amazing area!   
1  Downtown Harbourfront Private Room   
2      Seaton Village Parlour Bedroom   

                                             summary  \
0  This large, family home is located in one of T...   
1  Guest room in a luxury condo with access to al...   
2   Comfortable sofa bed in airy main floor parlour.   

                                               space  \
0  Visit Toronto with this fabulous furnished hom...   
1  My husband and I are both artists. I give pain...   
2  Family of 4 (2 kids)  plus a friendly dog open...   

                                         description experiences_offered  \
0  This large, family home is located in one of T...                none   
1  Guest room in a luxury condo with access to al...                none   
2  Comfortable sofa bed in airy main floor parlou...                none   

                               neighborhood_overview  \
0  The apartment is located in the Ossington stri...   
1                                                NaN   
2                                                NaN   

                                               notes  \
0  To learn a little more about our home, see: (U...   
1                                                NaN   
2                                                NaN   

                                             transit  \
0  All forms of public transportation are easily ...   
1                                                NaN   
2                                                NaN   

                                      access  \
0  The whole house and garden are available.   
1                                        NaN   
2                                        NaN   

                                         interaction  \
0  We will be out of town, but fully accessible b...   
1                                                NaN   
2                                                NaN   

                                         house_rules  thumbnail_url  \
0  This is our much-loved family home. We expect ...            NaN   
1                                 No smoking please.            NaN   
2                                         no smoking            NaN   

   medium_url                                        picture_url  \
0         NaN  https://a0.muscache.com/im/pictures/76206750/d...   
1         NaN  https://a0.muscache.com/im/pictures/11780344/1...   
2         NaN  https://a0.muscache.com/im/pictures/5282025/91...   

   xl_picture_url  host_id                                 host_url  \
0             NaN     1565   https://www.airbnb.com/users/show/1565   
1             NaN    22795  https://www.airbnb.com/users/show/22795   
2             NaN    48239  https://www.airbnb.com/users/show/48239   

        host_name host_since             host_location  \
0       Alexandra 2008-08-08  Toronto, Ontario, Canada   
1  Kathie & Larry 2009-06-22  Toronto, Ontario, Canada   
2            Rona 2009-10-25  Toronto, Ontario, Canada   

                                          host_about  host_response_time  \
0  I live in Toronto, Canada with my husband and ...                 NaN   
1  My husband and I have been airbnb host for alm...  a few days or more   
2  I love travelling and exploring new places, ma...                 NaN   

   host_response_rate  host_acceptance_rate host_is_superhost  \
0                 NaN                   NaN                 f   
1                 0.0                   NaN                 f   
2                 NaN                   NaN                 f   

                                  host_thu

In [25]:
df.neighbourhood.value_counts()

Downtown Toronto           4830
Entertainment District     1154
Niagara                     914
Willowdale                  707
The Annex                   671
Trinity-Bellwoods           495
Parkdale                    490
Riverdale                   460
Palmerston/Little Italy     362
Roncesvalles                360
Newtonbrook                 339
Little Portugal             312
Davisville                  310
Dufferin Grove              290
Fairbank                    272
Wallace Emerson             265
Dovercourt Park             263
Mimico                      245
York University Heights     228
Harbourfront                226
The Beaches                 202
Islington                   180
High Park North             178
Garden District             178
Fashion District            176
Glen Park                   172
L'Amoreaux                  168
Oakwood                     160
Danforth Village            153
Yorkville                   147
                           ... 
Scarboro

Overall, 54 of the 106 features consist of missing values, with missing counts from 1 (0.005%) to 20769 (100%). Features with excessive missing values are considered unuseful and will not be used for further analysis.

## 3.2 Data Type Correction 

### 3.2.1 Change Feature Data Type To Numeric

In the dataset, the features price, security_deposit, cleaning_fee, extra_people and host_response_rate are in string format. It is necessary to convert their data type to numeric. 

In [13]:
col_to_numeric = ['price','security_deposit','cleaning_fee','extra_people']
for col in col_to_numeric:
    df[col] = df[col].str.replace('$','').str.replace(',','').astype(float)
df.host_response_rate = df.host_response_rate.str.replace('%','').astype(float)/100

### 3.2.2 Change Feature Data Type to Datetime

In the dataset, the features last_scraped,host_since,calendar_last_scraped,first_review and last_review are in string format. It is necessary to convert their data type to numeric.

In [14]:
col_to_datetime = ['last_scraped','host_since','calendar_last_scraped','first_review','last_review']
for col in col_to_datetime:
    df[col] = pd.to_datetime(df[col])

## 3.X Extreme Values Identification

Since price is the target variable, the data integrity of this feature is of utmost importance. In this section we will evaluate the data is reliable, particularly for those listings with high values. 

In [15]:
df.price.describe()

count    20769.000000
mean       143.288074
std        234.157135
min          0.000000
25%         64.000000
50%        101.000000
75%        160.000000
max      13422.000000
Name: price, dtype: float64

In [17]:
df.loc[(df.price == 0),'listing_url']

8691    https://www.airbnb.com/rooms/20235002
8760    https://www.airbnb.com/rooms/20350463
8889    https://www.airbnb.com/rooms/20579837
9030    https://www.airbnb.com/rooms/20715061
Name: listing_url, dtype: object

While the median price is \$101, the highest nightly price in the dataset is \$13,422, and there are 26 listings with nightly price of \$2,000 or higher. Since those high prices are likely to affect our subsequent data analysis, we visit the url's of those listings to ensure the prices are reliable.   

## 3.3 Missing Values Imputation

For this section, we will only focus on features that will likely be relevent to data analysis.  
For imputation of each feature, we will create a new column with name 'variable_NA' to identify the rows that originally consists of the missing value. It may be useful information if the reason for missing is systemic (i.e. non-random).  

### 3.3.1 Numeric Features

Various numeric features consist of missing values. They will be grouped by the method of imputation.  
For the features host_listings_count, host response rate, bathrooms, bedrooms, beds, the missing values will be imputed with the feature median values.

In [41]:
for col in ['host_listings_count','bathrooms','host_response_rate','bedrooms','beds']:
    df[col+'_NA'] = np.where(df[col].isnull(),True, False)
    m = df[col].median()
    df[col].fillna(m,inplace=True)

Missing security deposits and cleaning fees are likely due the hosts' decision to not include one, which is equivalent to a value of 0. As such, the missing values will be imputed with 0.

In [43]:
for col in ['security_deposit','cleaning_fee']:
    df[col+'_NA'] = np.where(df[col].isnull(),True, False)
    df[col].fillna(0,inplace=True)

For review scores, the missing values are likely due to the facts that either the listings are new with few customers, or their customers did not leave a review score. The missing values are imputed with the feature median values.

In [44]:
for col in ['review_scores_rating','review_scores_accuracy','review_scores_cleanliness','review_scores_checkin',
           'review_scores_communication','review_scores_location','review_scores_value','reviews_per_month']:
    df[col+'_NA'] = np.where(df[col].isnull(),True, False)
    df[col].fillna(df[col].median(),inplace=True)

### 3.3.2 Categorical Features

The only categorical feature of interest with missing value is host_response_time. It is imputed with the feature mode.

In [54]:
df['host_response_time'+"_NA"] = np.where(df[col].isnull(),1,0)
df['host_response_time'].fillna(df['host_response_time'].mode().iloc[0],inplace=True)

### 3.3.3 Datetime Features

The datetime features with missing values are host_since, first_review, and last_review. The missing values are imputed with feature medians.

In [56]:
for col in ['host_since', 'first_review', 'last_review']:
    df[col+"_NA"] = np.where(df[col].isnull(),1,0)
    df[col].fillna(df[col].quantile(0.5),inplace=True)

## 3.4 New Feature Creation 

### 3.4.1 Dummy Variable for Amenities

The feature amenities contains a list of attributes provided by the host that the property contains. It is stored as a string enclosed in curly brackets with each attribute separated by comma. To further evaluate those amenities, a dummy variable is created for each amenity, with '1' and '0' indicating the presence and absence of that amenity, respectively. 

In [57]:
df['amenities_list'] = df['amenities'].str.lower().str.replace('[{}"]','').str.split(',')
amenities_set = set()
for i in df['amenities_list']:
    amenities_set.update(i)
amenities_set.remove('')
for item in amenities_set:
    df[item] = df.amenities_list.apply(lambda x: 1 if item in x else 0) 

There are 196 unique amenities.

## 3.4.2 City Names

One way to identify the geographical location of each listing is to map the city they are located in through its zip code. The first 3 characters of Canada's postal code, known as the Forward Sortation Area (FSA), will be used.  

In [85]:
df['fsa'] = df.zipcode.str.slice(stop=3).str.upper()
toronto_fsa = pd.read_csv('data/toronto_fsa.csv')
fsa_dict = {i:j for i,j in zip(toronto_fsa.Postcode, toronto_fsa.Borough)}
df['city_fsa'] = np.nan
df['city_fsa'] = df.fsa.map(fsa_dict)

A count of listings for each Toronto region is shown in the following list.

In [107]:
df.city_fsa.value_counts(dropna=False)

Downtown Toronto    7938
West Toronto        3361
North York          2749
Central Toronto     1426
Scarborough         1359
Etobicoke           1083
East Toronto        1069
York                 707
East York            671
NaN                  376
Thornhill             20
Not assigned           2
Mississauga            2
Pickering              2
Kleinburg              1
Guelph                 1
Queen's Park           1
Markham                1
Name: city_fsa, dtype: int64

There are three issues:
1. One listing is in Guelph. Since it is not in the Greater Toronto Area region, the listing will be removed.
2. The city is missing for 376 listings. It is due to missing FSA.
3. The city is not assigned for 2 listings. It is due to the fact that the FSA is not assigned to any cities, which means the provided FSA is likely inaccurate.  

For (2) and (3), we will use the provided longitude and latitude information, and [Google Map Platform](https://developers.google.com/maps/documentation/javascript/examples/geocoding-reverse) to reverse geocode, i.e. to use the longitude and latitude to obtain the postal codes. 

In [115]:
df.loc[df.city_fsa == "Queen's Park",['listing_url','longitude','latitude','price','zipcode', 'fsa']]

listing_url  longitude  latitude  price  zipcode  \
86  https://www.airbnb.com/rooms/289239  -79.41038  43.64294  199.0  M7A 2A9   

    fsa  
86  M7A

In [108]:
df_fsa_nan = df.loc[df.city_fsa.isnull(),['fsa','longitude','latitude']]
df_fsa_nan['lat-lon'] = df['latitude'].map(str) + ', '+ df['longitude'].map(str)
df_fsa_nan.to_csv('data/df_fsa_nan.csv')
# A new file is with the reverse geocoded information (address) is created 
df_rev_geocode1 = pd.read_csv('data/df_fsa_nan_address.csv')
df_rev_geocode1['fsa'] = df_rev_geocode1['address'].re('[A-Z][0-9][A-Z]')

In [109]:
df_rev_geocode1['fsa'] = df_rev_geocode1['address'].re('[A-Z][0-9][A-Z]')

AttributeError: 'Series' object has no attribute 're'

In [117]:
df[['latitude','longitude']].to_csv('data/lat_lon.csv', index=False)

In [118]:
df.head()

id                         listing_url       scrape_id last_scraped  \
0   1419   https://www.airbnb.com/rooms/1419  20190604202502   2019-06-05   
1   8077   https://www.airbnb.com/rooms/8077  20190604202502   2019-06-06   
2  12604  https://www.airbnb.com/rooms/12604  20190604202502   2019-06-05   
3  23691  https://www.airbnb.com/rooms/23691  20190604202502   2019-06-06   
4  26654  https://www.airbnb.com/rooms/26654  20190604202502   2019-06-05   

                                              name  \
0                  Beautiful home in amazing area!   
1               Downtown Harbourfront Private Room   
2                   Seaton Village Parlour Bedroom   
3                  Queen Bedroom close to downtown   
4  World Class downtown@ CN Tower! Jays, TIFF, ACC   

                                             summary  \
0  This large, family home is located in one of T...   
1  Guest room in a luxury condo with access to al...   
2   Comfortable sofa bed in airy main floor parlour.   
3                                                NaN   
4  CN Tower, TIFF Bell Lightbox, Metro Convention...   

                                               space  \
0  Visit Toronto with this fabulous furnished hom...   
1  My husband and I are both artists. I give pain...   
2  Family of 4 (2 kids)  plus a friendly dog open...   
3  Nestled in the middle of a quiet residential n...   
4  Welcome to 5***** Living - Bohemian decor~ in ...   

                                         description experiences_offered  \
0  This large, family home is located in one of T...                none   
1  Guest room in a luxury condo with access to al...                none   
2  Comfortable sofa bed in airy main floor parlou...                none   
3  Nestled in the middle of a quiet residential n...                none   
4  CN Tower, TIFF Bell Lightbox, Metro Convention...                none   

                               neighborhood_overview  \
0  The apartment is located in the Ossington stri...   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                               notes  \
0  To learn a little more about our home, see: (U...   
1                                                NaN   
2                                                NaN   
3  * Laundry facilities are available at the prem...   
4                                                NaN   

                                             transit  \
0  All forms of public transportation are easily ...   
1                                                NaN   
2                                                NaN   
3  * Public transit is conveniently located about...   
4          Walk everywhere! Transit at your doorstep   

                                      access  \
0  The whole house and garden are available.   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

                                         interaction  \
0  We will be out of town, but fully accessible b...   
1                                                NaN   
2                                                NaN   
3  You are welcome to contact us at any time if t...   
4                                                NaN   

                                         house_rules  thumbnail_url  \
0  This is our much-loved family home. We expect ...            NaN   
1                                 No smoking please.            NaN   
2                                         no smoking            NaN   
3  * Guests are requested to leave their shoes at...            NaN   
4  No Smoking Parking Available on Request ($20 d...            NaN   

   medium_url          

In [105]:
df.loc[df.city_fsa.isnull(),'is_location_exact'].value_counts()

f    343
t     33
Name: is_location_exact, dtype: int64